frontend/ui/widget/progressbardialog.lua

```diff
@@ -0,0 +1,191 @@
+local UIManager       = require("ui/uimanager")
+local Device          = require("device")
+local Screen          = Device.screen
+local VerticalGroup   = require("ui/widget/verticalgroup")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local Size            = require("ui/size")
+local FrameContainer  = require("ui/widget/container/framecontainer")
+local Blitbuffer      = require("ffi/blitbuffer")
+local ProgressWidget  = require("ui/widget/progresswidget")
+local dbg             = require("dbg")
+local TextWidget      = require("ui/widget/textwidget")
+local Font            = require("ui/font")
+local time            = require("ui/time")
+
+--[[--
+A dialog that shows a progress bar with a title and subtitle
+
+@usage
+local progressbar_dialog = ProgressbarDialog:new {
+    title = nil,
+    subtitle = nil,
+    progress_max = nil
+    refresh_time_seconds = 3,
+}
+
+-- attach progress callback and call show()
+progressbar_dialog:show()
+
+-- call close() when download is done
+progressbar_dialog:close()
+
+-----------------------general use case-----------------------------------------
+-- manually call reportProgress with the current progress (value between 0-progress_max)
+progressbar_dialog:reportProgress( <progress value> )
+----------------------------------------------------------------------------------
+
+------------------use case with luasocket sink----------------------------------
+local sink = ltn12.sink.file(io.open(local_path, "w"))
+sink = socketutil.chainSinkWithProgressCallback(sink, function(progress)
+    progressbar_dialog:reportProgress(progress)
+end)
+-- start pushing data to the sink, this is usually some function that accepts a luasocket sink
+something:startDownload(sink)
+----------------------------------------------------------------------------------
+
+Note: provide at least one of title, subtitle or progress_max
+@param title string the title of the dialog
+@param subtitle string the subtitle of the dialog
+@param progress_max number the maximum progress (e.g. size of the file in bytes for file downloads)
+                    reportProgress() should be called with the current
+                    progress (value between 0-progress_max) to update the progress bar
+                    optional: if `progress_max` is nil, the progress bar will be hidden
+@param refresh_time_seconds number refresh time in seconds
+--]]
+local ProgressbarDialog = WidgetContainer:extend {}
+
+function ProgressbarDialog:init()
+    self.align = "center"
+    self.dimen = Screen:getSize()
+
+    -- set default values if not provided
+    self.title = self.title ~= nil and self.title or nil
```

Comment:
I don't get, what does this line do?

Reply:
I had a different default value while testing and ended up changing the default to nil anyway. 

I left it in there to show the intention that title, subtile and progress_max was nil-able but at this point it should can be obvious so this can be removed, the comments and usage make this clear enough